filebeat/input/mqtt/client.go

```diff
@@ -0,0 +1,207 @@
+
+		// Certificates = list of certs client sends to server.
+		tlsconfig.Certificates = []tls.Certificate{cert}
+	}
+
+	// Create tls.Config with desired tls properties
+	return tlsconfig, nil
+}
+
+// Prepare MQTT client
+func (input *mqttInput) setupMqttClient() error {
+	c := input.config
+
+	logp.Info("[MQTT] Connect to broker URL: %s", c.Host)
+
+	mqttClientOpt := MQTT.NewClientOptions()
+	mqttClientOpt.SetClientID(c.ClientID)
+	mqttClientOpt.AddBroker(c.Host)
+
+	mqttClientOpt.SetMaxReconnectInterval(1 * time.Second)
+	mqttClientOpt.SetConnectionLostHandler(input.connectionLostHandler)
+	mqttClientOpt.SetOnConnectHandler(input.subscribeOnConnect)
+	mqttClientOpt.SetAutoReconnect(true)
+
+	if c.Username != "" {
+		logp.Info("[MQTT] Broker username: %s", c.Username)
+		mqttClientOpt.SetUsername(c.Username)
+	}
+
+	if c.Password != "" {
+		mqttClientOpt.SetPassword(c.Password)
+	}
+
+	if c.SSL == true {
+		logp.Info("[MQTT] Configure session to use SSL")
+		tlsconfig, err := input.newTLSConfig()
+		if err != nil {
+			return err
+		}
+		mqttClientOpt.SetTLSConfig(tlsconfig)
+	}
+
+	input.client = MQTT.NewClient(mqttClientOpt)
+	return nil
+}
+
+func (input *mqttInput) connect() error {
+	if token := input.client.Connect(); token.WaitTimeout(input.config.WaitClose) && token.Error() != nil {
+		logp.Err("MQTT Failed to connect")
+		return token.Error()
+	}
+	logp.Info("MQTT Client connected: %t", input.client.IsConnected())
+	return nil
+}
+
+func (input *mqttInput) subscribeOnConnect(client MQTT.Client) {
+	subscriptions := prepareSubscriptionsForTopics(input.config.Topics, input.config.QoS)
+
+	// Mqtt client - Subscribe to every topic in the config file, and bind with message handler
+	if token := input.client.SubscribeMultiple(subscriptions, input.onMessage); token.WaitTimeout(input.config.WaitClose) && token.Error() != nil {
+		logp.Error(token.Error())
+	}
+	logp.Info("MQTT Subscribed to configured topics")
+}
+
+// Mqtt message handler
+func (input *mqttInput) onMessage(client MQTT.Client, msg MQTT.Message) {
+	logp.Debug("MQTT", "MQTT message received: %s", string(msg.Payload()))
+	var beatEvent beat.Event
+	eventFields := make(common.MapStr)
+
+	// default case
+	var mqtt = make(common.MapStr)
+	eventFields["message"] = string(msg.Payload())
+	if input.config.DecodePayload {
+		mqtt["fields"] = decodeBytes(msg.Payload())
+	}
+
+	eventFields["is_system_topic"] = strings.HasPrefix(msg.Topic(), "$")
+	eventFields["topic"] = msg.Topic()
+
+	mqtt["id"] = msg.MessageID()
+	mqtt["retained"] = msg.Retained()
+	eventFields["mqtt"] = mqtt
+
+	// Finally sending the message to elasticsearch
+	beatEvent.Fields = eventFields
+	input.outlet.OnEvent(beatEvent)
+
+	logp.Debug("MQTT", "Event sent: %t")
+}
+
+// connectionLostHandler will try to reconnect when connection is lost
+func (input *mqttInput) connectionLostHandler(client MQTT.Client, reason error) {
+	logp.Warn("[MQTT] Connection lost: %s", reason.Error())
+
+	//Rerun the input
+	input.Run()
```

Comment:
This won't do anything, because of the once guard in `Run`... that once is because the calling code might call `Run` multiple times, which is the right behavior for the log input but not most of the others. If this is the right program flow (which seems plausible) then you should split `Run` into public and internal pieces, with `run` doing the actual logic and `Run` just protecting it with a `Once`. Then the internal version can be called here.